### DISCUSSION TITLE
List of all possible widgets we can theme in an LXQt theme?

### DISCUSSION THREAD

**Discussion by AzumaHazuki** [lxqt/lxqt#2354] [Q&A]
This isn't the first time I've asked this, but it seems to have been a bit ambiguous last time, so I'll try again:

Is there, anywhere on the internet, a comprehensive list of all the widgets, subcontrols, and pseudo-states thereof *specific to LXQt and its themes* that may be styled in our QSS stylesheets (lxqt-panel.qss etc.)? I ask because every time I look at someone else's work there are theming directives I had no idea existed; <a href=https://github.com/pixelocdguy/windows-classic/blob/main/lxqt/windows-classic/lxqt-panel.qss>PixelDocGuy's WinClassic theme</a> taught me how to create a menu banner, for example, and so far is the only instance I've ever seen of #TopLevelMainMenu (and also some clever margins trickery to enable the banner...devious!).

We really, really need this. Tsujan raised a concern about themes being inconsistent or incomplete in an earlier thread, and having this would go a long way toward making sure this doesn't become an issue.

**Comment by tsujan** [comment]
Yes, a theme-making wiki is missing. We lack the required manpower. Contributions are welcome.

For now, there are only two ways for finding stylable LXQt widgets: (1) The direct way, by consulting the code; and (2) trial and error, by consulting the existing themes.

**Reply by palinek** [reply]
And (3) reviewing the objects hierarchy in panel's DOM plugin.

**Reply by tsujan** [reply]
> reviewing the objects hierarchy in panel's DOM plugin.

I totally missed it. Thanks!

It needs to be turned on with compilation; it's off by default.

**Comment by M4he** [comment]
From what I've gathered so far, there are at least 3 different ways of addressing widgets in LXQt themes (ordered from generic to specific):

1. generic Qt5 widget classes like `QWidget`, `QToolButton` etc. as per https://doc.qt.io/qt-5/widget-classes.html (LXQt uses only a specific subset of those)
2. custom classes implemented by LXQt, mostly derived from basic Qt5 widgets, for example `VolumePopup` based on `QDialog`, see [lxqt/lxqt-panel/plugin-volume/volumepopup.h](https://github.com/lxqt/lxqt-panel/blob/3bfd7b9ea3a844d85a995f91451e88cd0eb602df/plugin-volume/volumepopup.h#L37)
3. specific widget identifiers set by `themeId` definitions (do not exist for every widget), for example in lxqt-panel: https://github.com/lxqt/lxqt-panel/search?q=themeid

Cases 1 and 2 can be addressed as classes in the theme QSS (including hierarchies):

```css
/* address any QWidget */
QWidget {
    ...
}

/* address push buttons only directly within the VolumePopup widget class */
VolumePopup  > QPushButton {
    ...
}
```

Case 3 is the most specific and requires a `#` in front of it because it is an identifier, not a class:

```css
#TaskBar {
    ...
}
```

More specific declarations will override generic ones, i.e. if you only define `QWidget` styles, it will apply to many different widgets. However, if you address a `#` identifier belonging to a `QWidget`, it will use the specific values declared in the identifier's section instead of the more generic `QWidget {...}` style (although it keeps values from its broader class whenever they aren't explicitly set in the more specific one). Same goes for hierarchies - QSS is like CSS in this regard, see https://doc.qt.io/qt-5/stylesheet-syntax.html

Gathering all possible occurences for 1 and 3 should be feasible. Gathering all custom classes for case 2 is more manual code digging I guess.

**Reply by tsujan** [reply]
@M4he, thanks!

> Gathering all possible occurences for 1 and 3 should be feasible

As you correctly said, 1 is in the Qt doc. We shouldn't duplicate it.

A Wiki will need a complete list of 2 and 3. Also, a concise explanation like yours could be included. The rest is the Qt doc, whose link should be added to the Wiki, of course.

**Comment by AzumaHazuki** [comment]
This is excellent progress on the subject :) @M4he, thank you in particular for pointing out the themeID strings, as these appear to point to most if not all of the custom widgets LXQt uses.

There is a block of QSS/CSS selectors at the start of most themes that always looked a bit cargo-cult-ish to me, and I'm not sure what precisely it's doing: it is *extremely* specific, in that it's using the > ("next-level descendant only") operator, sometimes in rapid succession. That looks almost surgically targeted and absolutely intentional:

```
Plugin > QWidget,
Plugin > QWidget > QWidget {
...
}

Plugin > QToolButton,
Plugin > QWidget > QToolButton,
Plugin > QWidget > QToolButton > QWidget > QToolButton,
LXQtPanelPlugin > QToolButton {
...
}

```
So is this a specially-crafted selector cohort that catches all widgets, and *only* those widgets, that LXQt themes are supposed to reach? Should this be part of all themes aiming for a consistent look? 

I do know that using `QWidget { ... }` or even, yikes, `* { ... }` will affect things one _really_ does not expect to affect, like the customization dlalogs for the panel and notifications. Is this because they're children of the panel?
